pkg/cgroups/manager/manager.go

```diff
@@ -0,0 +1,392 @@
+
+	return m
+}
+
+func (m *CgroupManager) enable() {
+	if !option.Config.EnableSocketLBTracing {
+		m.enabled = false
+		return
+	}
+	m.enabled = true
+	m.checkCgroupPath.Do(func() {
+		for _, path := range cgroupBasePaths {
+			if _, err := m.fschecker.Stat(path); err != nil {
+				continue
+			}
+			m.templateCgroupBasePath = path
+			break
+		}
+		if m.templateCgroupBasePath == "" {
+			log.Warnf("No valid cgroup base path found: socket " +
+				"load-balancing tracing feature will not work")
+			m.enabled = false
+		}
+	})
+
+	if m.enabled {
+		log.Info("Cgroup metadata manager is enabled")
+	}
+}
+
+func (m *CgroupManager) processPodEvents() {
+	for {
+		select {
+		case ev := <-m.podEvents:
+			if !m.enabled {
+				continue
+			}
+			switch ev.eventType {
+			case podAddEvent, podUpdateEvent:
+				m.updatePodMetadata(ev.pod, ev.oldPod)
+			case podDeleteEvent:
+				m.deletePodMetadata(ev.pod)
+			case podGetMetadataEvent:
+				m.getParentPodMetadata(ev.cgroupId, ev.podMetadataOut)
+			}
+		case <-m.shutdown:
+			return
+		}
+	}
+}
+
+func (m *CgroupManager) updatePodMetadata(pod, oldPod *v1.Pod) {
+	id := string(pod.ObjectMeta.UID)
+	pm, ok := m.podMetadataById[id]
+	if !ok {
+		// Fill in pod static metadata.
+		pm = &podMetadata{
+			name:      pod.Name,
+			namespace: pod.Namespace,
+		}
+		m.podMetadataById[id] = pm
+	}
+	if oldPod != nil && oldPod.Status.DeepEqual(&pod.Status) {
+		return
+	}
+	// Only update the metadata that can change. This excludes pod's name,
+	// namespace, id, and qos class.
+	podIPs := pod.Status.PodIPs
+	pm.ips = make([]string, len(podIPs))
+	for i := range podIPs {
+		pm.ips[i] = podIPs[i].IP
+	}
+	// Get metadata for pod's containers that are in the running state. Containers
+	// can get re-created, and their ids can change. Update the new containers.
+	// Pod's metadata including its containers map will be deleted when the pod
+	// is deleted.
+	numContainers := len(pod.Status.ContainerStatuses)
+	if pm.containers == nil && numContainers > 0 {
+		pm.containers = make(map[string]struct{})
+	}
+	currContainers := make(map[string]struct{}, numContainers)
+	for _, c := range pod.Status.ContainerStatuses {
+		var cId string
+		if cId = c.ContainerID; cId == "" || c.State.Running == nil {
+			continue
+		}
+		// The container ID field is of the form: <container-runtime>://<containerID>
+		// Example:containerd://e275d1a37782ab30008aa3ae6666cccefe53b3a14a2ab5a8dc459939107c8c0e
+		_, after, found := strings.Cut(cId, "//")
+		if !found || after == "" {
+			log.WithFields(logrus.Fields{
+				logfields.K8sPodName:   pod.Name,
+				logfields.K8sNamespace: pod.Namespace,
+			}).Errorf("unexpected container ID: %s", cId)
+			continue
+		}
+		cId = after
+		if _, ok := pm.containers[cId]; ok {
+			currContainers[cId] = struct{}{}
+			// Container cgroup path doesn't change as long as the container id
+			// is the same.
+			continue
+		}
+		pm.containers[cId] = struct{}{}
+		currContainers[cId] = struct{}{}
```

Comment:
Should these wait until after the `continue` statements below since we could end up skipping this container if we can't locate the cgroup path or id from that path?

Reply:
Cgroup path or id don't change for a container. So my thinking is that we don't want to repeat the processing. Wdyt?

Reply:
This question is maybe a bit more theoretical: broadly there are "error" cases in the "continue" statements below which could cause this container to be skipped. In that case, `m.podMetadataById[id]` would be updated via `pm.containers[cId]` here, but this loop wouldn't reach the `m.containerMetadataByCgrpId[cgrpId]` update below on line 323. Just as a consistency thing, I would think that the structure should ideally have both of these updated or neither.

Reply:
>In that case, m.podMetadataById[id] would be updated via pm.containers[cId] here, but this loop wouldn't reach the m.containerMetadataByCgrpId[cgrpId] update below on line 323. 

The error condition is that we weren't able to compute metadata for pod's containers. This can be useful for debugging when we add a CLI in future.